mountpoint-s3/src/prefetch/backpressure_controller.rs

```diff
@@ -91,58 +115,125 @@ impl BackpressureController {
-                    trace!(
-                        preferred_read_window_size = self.preferred_read_window_size,
-                        next_read_offset,
-                        read_window_end_offset = self.read_window_end_offset,
-                        to_increase,
-                        "incrementing read window"
-                    );
-                    self.increment_read_window(to_increase).await;
-                    self.read_window_end_offset = new_read_window_end_offset;
+
+                    // Force incrementing read window regardless of available memory when we are already at minimum
+                    // read window size.
+                    if self.preferred_read_window_size <= self.min_read_window_size {
+                        self.mem_limiter.reserve(to_increase as u64);
+                        self.increment_read_window(to_increase).await;
+                        break;
+                    }
+
+                    // Try to reserve the memory for the length we want to increase before sending the request,
+                    // scale down the read window if it fails.
+                    if self.mem_limiter.try_reserve(to_increase as u64) {
+                        self.increment_read_window(to_increase).await;
+                    } else {
+                        self.scale_down();
+                    }
                 }
             }
-            BackpressureFeedbackEvent::PartQueueStall => self.try_scaling_up(),
+            BackpressureFeedbackEvent::PartQueueStall => self.scale_up(),
         }
         Ok(())
     }
 
     // Send an increment read window request to the stream producer
-    async fn increment_read_window(&self, len: usize) {
+    async fn increment_read_window(&mut self, len: usize) {
+        trace!(
+            preferred_read_window_size = self.preferred_read_window_size,
+            next_read_offset = self.next_read_offset,
+            read_window_end_offset = self.read_window_end_offset,
+            len,
+            "incrementing read window"
+        );
+
         // This should not block since the channel is unbounded
         let _ = self
             .read_window_updater
             .send(len)
             .await
             .inspect_err(|_| trace!("read window incrementing queue is already closed"));
+        self.read_window_end_offset += len as u64;
     }
 
-    // Try scaling up preferred read window size with a multiplier configured at initialization.
-    fn try_scaling_up(&mut self) {
+    // Scale up preferred read window size with a multiplier configured at initialization.
+    // Scaling up fails silently if there is no enough free memory to perform it.
+    fn scale_up(&mut self) {
         if self.preferred_read_window_size < self.max_read_window_size {
+            let new_read_window_size = self.preferred_read_window_size * self.read_window_size_multiplier;
+            // Also align the new read window size to the client part size
```

Comment:
it took me a while to understand, why we're doing that; it's because CRT will round up the `read_window_end` to this value, consuming more memory, right? can we document this?

Reply:
Actually, I would just drop this because we now use allocated memory stat from the CRT (which is more than the actual mem used) and it doesn't really matter if we round up our window to the part size.

Reply:
But the excess will end up in our `PartQueue` and will not be reported by CRT though after some point, right? Probably until `part_size` is not huge this overhead should be negligible. I am wondering how bad it'll get if `part_size` is, say, 512mb, and unaccounted data stays up in the queue.

May be we can go on with dropping this alignment for now, but add a metric in `read_from_request` counting the excess (`backpressure_limiter.read_window_end_offset - request.next().await.offset`)?

Reply:
That's true if the part size were large then we would end up consuming much more memory than reserved.